sequencer/src/api.rs

```diff
@@ -3389,6 +3393,221 @@ mod test {
+                        "reward-state/reward-balance/{block}/{address}"
+                    ))
+                    .send()
+                    .await
+                    .ok()
+                    .flatten();
+                assert!(amount.is_none(), "amount is not none for block {block}")
+            }
+        }
+
+        // Collect leaves for epoch 3 to 5 to verify reward calculations.
+        let leaves = client
+            .socket("availability/stream/leaves/41")
+            .subscribe::<LeafQueryData<SeqTypes>>()
+            .await
+            .unwrap()
+            .take((EPOCH_HEIGHT * 3).try_into().unwrap())
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
+
+        let node_state = network.server.node_state();
+        let coordinator = node_state.coordinator;
+
+        let mut rewards_map = HashMap::new();
+
+        for leaf in leaves {
+            let block = leaf.height();
+            tracing::info!("verify rewards for block={block:?}");
+            let membership = coordinator.membership().read().await;
+            let epoch = epoch_from_block_number(block, EPOCH_HEIGHT);
+            let epoch_number = EpochNumber::new(epoch);
+            let leader = membership
+                .leader(leaf.leaf().view_number(), Some(epoch_number))
+                .expect("leader");
+            let leader_eth_address = membership.address(&epoch_number, leader).expect("address");
+            drop(membership);
+
+            let validators = client
+                .get::<IndexMap<Address, Validator<BLSPubKey>>>(&format!("node/validators/{epoch}"))
+                .send()
+                .await
+                .expect("validators");
+
+            let leader_validator = validators
+                .get(&leader_eth_address)
+                .expect("leader not found");
+
+            // Verify that the sum of delegator stakes equals the validator's total stake.
+            for validator in validators.values() {
+                let delegator_stake_sum: U256 = validator.delegators.values().cloned().sum();
+
+                assert_eq!(delegator_stake_sum, validator.stake);
+            }
+
+            let computed_rewards =
+                compute_rewards(leader_validator.clone()).expect("reward computation");
+
+            // Verify that the leader commission amount is within the tolerated range.
+            // Due to potential rounding errors in decimal calculations for delegator rewards,
+            // the actual distributed commission
+            // amount may differ very slightly from the calculated value.
+            // this asserts that it is within 10wei tolerance level.
+            // 10 wei is 10* 10E-18
+            let total_reward = block_reward().0;
+            let leader_commission_basis_points = U256::from(leader_validator.commission);
+            let calculated_leader_reward_amount = leader_commission_basis_points
```

Comment:
@imabdulbasit I guess I adapted the terminology from earlier reviews. Here we use `leader_reward` to mean the reward from commission only.

Reply:
https://github.com/EspressoSystems/espresso-network/pull/3249/commits/315c405c2747a6771cc0101ebd4e7ca3286f81e4